File: src/Butil/Bit.Butil/Publics/NetworkInformation.cs
using System;
using System.Collections.Concurrent;
using System.Diagnostics.CodeAnalysis;
using System.Threading.Tasks;
using Microsoft.JSInterop;

namespace Bit.Butil;

/// <summary>
/// Wraps the <see href="https://developer.mozilla.org/en-US/docs/Web/API/NetworkInformation">Network Information API</see>
/// (<c>navigator.connection</c>) plus the always-available <c>navigator.onLine</c>.
/// </summary>
public class NetworkInformation(IJSRuntime js) : IAsyncDisposable
{
    internal const string ChangeMethodName = nameof(InvokeNetworkChange);

    private readonly ConcurrentDictionary<Guid, Action<NetworkConnectionStatus>> _handlers = new();

    // Per-instance callback reference (see Keyboard): listeners are isolated per circuit / WASM app
    // and released on disposal - no static state, no cross-circuit leak.
    private DotNetObjectReference<NetworkInformation>? _dotNetRef;
    private DotNetObjectReference<NetworkInformation> DotNetRef => DotNetObjectReferenceHelper.GetOrCreate(ref _dotNetRef, this);

    /// <summary>True when the runtime exposes <c>navigator.connection</c>.</summary>
    /// <remarks>
    /// This is the Chromium-only half of the API. <see cref="GetStatus"/> works regardless - on a
    /// browser that answers <c>false</c> here it still reports
    /// <see cref="NetworkConnectionStatus.Online"/> from <c>navigator.onLine</c> and leaves the
    /// connection-quality fields null - so treat this as "is the detail available", not as
    /// "may I call GetStatus".
    /// <br/>
    /// During prerender/SSR (no JS runtime) this returns <c>default</c> (e.g. <c>false</c>/<c>0</c>)
    /// rather than throwing, so the result can't be distinguished from a genuine value. If you
    /// branch on it, defer the read to <c>OnAfterRenderAsync</c>.
    /// </remarks>
    public ValueTask<bool> IsSupported() => js.Invoke<bool>("BitButil.networkInformation.isSupported");

    /// <summary>One-shot snapshot of the network state.</summary>
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(NetworkConnectionStatus))]
    public ValueTask<NetworkConnectionStatus> GetStatus()
        => js.Invoke<NetworkConnectionStatus>("BitButil.networkInformation.getStatus");

    /// <summary>Invoked from JS on each connectivity change. See <see cref="SubscribeChange"/>.</summary>
    [JSInvokable(ChangeMethodName)]
    public void InvokeNetworkChange(Guid id, NetworkConnectionStatus status)
    {
        if (_handlers.TryGetValue(id, out var handler)) handler.Invoke(status);
    }

    /// <summary>
    /// Watches connectivity and calls <paramref name="handler"/> with a fresh snapshot whenever it
    /// changes - going offline or online, or (on Chromium) the connection quality shifting.
    /// </summary>
    /// <param name="handler">Called with the new state on every change.</param>
    /// <returns>A subscription - dispose it to detach the listeners.</returns>
    /// <remarks>
    /// Attaches to the window's <c>online</c>/<c>offline</c> events as well as
    /// <c>navigator.connection</c>'s <c>change</c>, so this is useful on every browser - it just
    /// won't report quality changes where <see cref="IsSupported"/> is false.
    /// <br/>
    /// Going "online" only means the device has a network attached, not that anything is reachable;
    /// a captive portal still reports online. Treat it as a hint to retry, not a guarantee.
    /// </remarks>
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(NetworkConnectionStatus))]
    [DynamicDependency(nameof(InvokeNetworkChange), typeof(NetworkInformation))]
    public async ValueTask<ButilSubscription> SubscribeChange(Action<NetworkConnectionStatus> handler)
    {
        ArgumentNullException.ThrowIfNull(handler);

        var id = Guid.NewGuid();
        _handlers[id] = handler;
        await js.InvokeVoid("BitButil.networkInformation.subscribe", DotNetRef, id);

        return new ButilSubscription(id, async () =>
        {
            _handlers.TryRemove(id, out _);
            await js.InvokeVoid("BitButil.networkInformation.unsubscribe", id);
        });
    }

    /// <summary>
    /// On scope/circuit teardown, detaches any listener whose <see cref="ButilSubscription"/> was
    /// never disposed.
    /// </summary>
    public async ValueTask DisposeAsync()
    {
        try
        {
            _handlers.Clear();
            await js.InvokeVoid("BitButil.networkInformation.disposeAll");
        }
        catch (Exception ex) when (ex.IsIgnorableDisposalException()) { } // teardown: circuit gone, cancelled, or already disposed
        finally
        {
            _dotNetRef?.Dispose();
            _dotNetRef = null;
        }
        GC.SuppressFinalize(this);
    }
}

File: src/Butil/Bit.Butil/Scripts/networkInformation.ts
var BitButil = BitButil || {};

(function (butil: any) {
    const _listeners: { [id: string]: { connection: any, handler: () => void } } = {};

    function connection() {
        const nav = window.navigator as any;
        return nav.connection || nav.mozConnection || nav.webkitConnection || null;
    }

    function snapshot() {
        const nav = window.navigator as any;
        const c = connection();
        return {
            online: !!nav.onLine,
            effectiveType: c?.effectiveType ?? null,
            type: c?.type ?? null,
            downlink: c?.downlink ?? null,
            downlinkMax: c?.downlinkMax ?? null,
            rtt: c?.rtt ?? null,
            saveData: typeof c?.saveData === 'boolean' ? c.saveData : null
        };
    }

    butil.networkInformation = {
        // navigator.onLine is everywhere; navigator.connection - the part that carries the
        // effective type, downlink and save-data - is Chromium only. GetStatus works either way
        // and reports nulls for the missing half, so this says which half you are getting.
        isSupported() { return !!connection(); },
        getStatus: snapshot,
        subscribe(dotNetRef: any, listenerId: string) {
            const handler = () => butil.utils.dispatch(dotNetRef, 'InvokeNetworkChange', listenerId, snapshot());
            const c = connection();

            // window's online/offline fire everywhere, so the subscription is useful even without
            // navigator.connection - it just won't report quality changes on those browsers.
            window.addEventListener('online', handler);
            window.addEventListener('offline', handler);
            c?.addEventListener('change', handler);

            _listeners[listenerId] = { connection: c, handler };
        },
        unsubscribe(listenerId: string) {
            const entry = _listeners[listenerId];
            if (!entry) return;
            delete _listeners[listenerId];
            window.removeEventListener('online', entry.handler);
            window.removeEventListener('offline', entry.handler);
            entry.connection?.removeEventListener('change', entry.handler);
        },
        disposeAll() {
            for (const id of Object.keys(_listeners)) butil.networkInformation.unsubscribe(id);
        }
    };
}(BitButil));
